exchanges/okx/okx_wrapper.go

```diff
@@ -0,0 +1,1218 @@
+			QuoteVolume:  mdata.VolCcy24H,
+			Open:         mdata.Open24H,
+			Pair:         p,
+			ExchangeName: ok.Name,
+			AssetType:    a,
+		})
+		if er != nil {
+			return nil, er
+		}
+	default:
+		return nil, fmt.Errorf("assetType not supported: %v", a)
+	}
+	return ticker.GetTicker(ok.Name, p, a)
+}
+
+// UpdateTickers updates all currency pairs of a given asset type
+func (ok *Okx) UpdateTickers(ctx context.Context, assetType asset.Item) error {
+	switch assetType {
+	case asset.Spot, asset.Margin, asset.Futures, asset.PerpetualSwap, asset.Option:
+		instrumentType := ok.GetInstrumentTypeFromAssetItem(assetType)
+		ticks, er := ok.GetTickers(ctx, strings.ToUpper(instrumentType), "", "")
+		if er != nil {
+			return er
+		}
+		pairs, er := ok.GetEnabledPairs(assetType)
+		if er != nil {
+			return er
+		}
+		for i := range pairs {
+			for y := range ticks {
+				pairFmt, err := ok.FormatExchangeCurrency(pairs[i], assetType)
+				if err != nil {
+					return err
+				}
+				pair, er := ok.GetPairFromInstrumentID(ticks[y].InstrumentID)
+				if er != nil {
+					return er
+				}
+				if pair.String() != pairFmt.String() {
+					continue
+				}
+				er = ticker.ProcessTicker(&ticker.Price{
+					Last:         ticks[y].LastTradePrice,
+					High:         ticks[y].High24H,
+					Low:          ticks[y].Low24H,
+					Bid:          ticks[y].BidPrice,
+					Ask:          ticks[y].BestAskPrice,
+					Volume:       ticks[y].Vol24H,
+					QuoteVolume:  ticks[y].VolCcy24H,
+					Open:         ticks[y].Open24H,
+					Pair:         pair,
+					ExchangeName: ok.Name,
+					AssetType:    assetType,
+				})
+				if er != nil {
+					return er
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// FetchTicker returns the ticker for a currency pair
+func (ok *Okx) FetchTicker(ctx context.Context, p currency.Pair, assetType asset.Item) (*ticker.Price, error) {
+	formatedPair, er := ok.FormatExchangeCurrency(p, assetType)
+	if er != nil {
+		return nil, er
+	}
+	tickerNew, err := ticker.GetTicker(ok.Name, formatedPair, assetType)
+	if err != nil {
+		return ok.UpdateTicker(ctx, p, assetType)
+	}
+	return tickerNew, nil
+}
+
+// FetchOrderbook returns orderbook base on the currency pair
+func (ok *Okx) FetchOrderbook(ctx context.Context, pair currency.Pair, assetType asset.Item) (*orderbook.Base, error) {
+	ob, err := orderbook.Get(ok.Name, pair, assetType)
+	if err != nil {
+		return ok.UpdateOrderbook(ctx, pair, assetType)
+	}
+	return ob, nil
+}
+
+// UpdateOrderbook updates and returns the orderbook for a currency pair
+func (ok *Okx) UpdateOrderbook(ctx context.Context, pair currency.Pair, assetType asset.Item) (*orderbook.Base, error) {
+	book := &orderbook.Base{
+		Exchange:        ok.Name,
+		Pair:            pair,
+		Asset:           assetType,
+		VerifyOrderbook: ok.CanVerifyOrderbook,
+	}
+
+	var orderbookNew *OrderBookResponse
+	var er error
+	switch assetType {
+	case asset.Spot, asset.Margin, asset.PerpetualSwap, asset.Option, asset.Futures:
```

Comment:
In these instances you can drop this switch case and use method:

```go
if !ok.SupportsAsset(a) {
		return fmt.Errorf("%w %s", asset.ErrNotSupported, a)
}
```
It reduces a lot of nesting. You can update across the wrapper file.